pkg/reporting/sink.go

```diff
@@ -3,21 +3,31 @@ package reporting
 // public API for event reporting
 
 import (
+	"context"
 	"fmt"
 	"sync"
 
 	"github.com/prometheus/client_golang/prometheus"
 	"go.uber.org/zap"
+
+	"github.com/neondatabase/autoscaling/pkg/util/taskgroup"
 )
 
 type EventSink[E any] struct {
 	queueWriters []eventQueuePusher[E]
 	done         func()
 }
 
-func NewEventSink[E any](logger *zap.Logger, metrics *EventSinkMetrics, clients ...Client[E]) *EventSink[E] {
+func NewEventSink[E any](
+	ctx context.Context,
+	logger *zap.Logger,
+	tg taskgroup.Group,
+	metrics *EventSinkMetrics,
+	clients ...Client[E],
+) *EventSink[E] {
 	var queueWriters []eventQueuePusher[E]
-	signalDone := make(chan struct{})
+
+	senderCtx, cancelSenders := context.WithCancel(ctx)
```

Comment:
It seems to me you don't really need `Finish()`, `done` and `senderCtx`. You can just pass `ctx` to `senderLoop` and have it the only way to stop sender loop.

Reply:
Resolving as this is no longer relevant because I changed the approach.